primitives/s2s/src/call_params.rs

```diff
@@ -0,0 +1,193 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// --- paritytech ---
+use bp_message_dispatch::CallOrigin;
+use bp_runtime::messages::DispatchFeePayment;
+use sp_core::{H160, U256};
+use sp_runtime::{AccountId32, MultiSignature, MultiSigner};
+use sp_std::{vec, vec::Vec};
+// --- darwinia-network ---
+use codec::{Decode, Encode};
+use dp_asset::token::TokenMetadata;
+
+/// The parameters box for the pallet runtime call.
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub enum CallParams {
+	#[codec(index = 0)]
+	S2sIssuingPalletRegisterFromRemote(TokenMetadata),
+	#[codec(index = 1)]
+	S2sIssuingPalletIssueFromRemote(H160, U256, H160),
+	#[codec(index = 2)]
+	S2sBackingPalletUnlockFromRemote(H160, U256, Vec<u8>),
+}
+/// Creating a concrete message payload which would be relay to target chain.
+pub trait CreatePayload<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature>
+where
+	SourceChainAccountId: Encode,
+	Self: Sized,
+{
+	type Payload: Sized + Encode;
+
+	fn encode_call(pallet_index: u8, call_params: CallParams) -> Result<Vec<u8>, &'static str> {
+		let mut encoded = vec![pallet_index];
+		encoded.append(&mut call_params.encode());
+		Ok(encoded)
+	}
+
+	fn create(
+		origin: CallOrigin<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature>,
+		spec_version: u32,
+		weight: u64,
+		call_params: CallParams,
+		dispatch_fee_payment: DispatchFeePayment,
+	) -> Result<Self::Payload, &'static str>;
+}
+
+impl CreatePayload<AccountId32, MultiSigner, MultiSignature> for () {
```

Comment:
do we need to implement this () type? where to use?

Reply:
Only in test now.